build.rs

```diff
@@ -85,6 +81,62 @@ fn build_and_link() -> IncludePaths {
     IncludePaths(vec![include])
 }
 
+fn autogen(build_dir: &Path) {
+    let mut command = Command::new("sh");
+
+    command
+        .arg("autogen.sh")
+        .current_dir(&build_dir);
+
+    if cfg!(windows) {
+        // When building in MSYS2, not clearing the environment variables first
+        // results in `configure` being generated incorrectly. By clearing the
+        // variables first, then restoring PATH, we can ensure the correct file
+        // is generated.
+        command
+            .env_clear()
+            .env("PATH", env::var("PATH").unwrap());
+    }
+
+    run_command("Generating configure", &mut command);
+}
+
+fn configure_libffi(prefix: PathBuf, build_dir: &Path) {
+    let mut command = Command::new("sh");
+
+    command.arg("configure")
+        .arg("--with-pic")
+        .current_dir(&build_dir);
+
+    if cfg!(windows) {
+        // When using MSYS2, OUT_DIR will be a Windows like path such as
+        // C:\foo\bar. Unfortunately, the various scripts used for building
+        // libffi do not like such a path, so we have to turn this into a Unix
+        // like path such as /c/foo/bar.
+        //
+        // This code assumes the path only uses : for the drive letter, and only
+        // uses \ as a component separator. It will likely break for file paths
+        // that include a :.
+        let mut msys_prefix = prefix
+            .to_str()
+            .unwrap()
+            .replace(":\\", "/")
+            .replace("\\", "/");
+
+        msys_prefix.insert(0, '/');
+
+        command
+            .arg("--prefix")
+            .arg(msys_prefix);
```

Comment:
This can be best described as "an ugly hack", but unfortunately I don't think there is an easier way of doing this. We can't use `PathBuf` here, since IIRC it would produce Windows paths (`C:\foo\bar`) instead of Unix paths (which we require). Using Windows paths in MSYS2 breaks the compilation process of libffi.